circle-mlir/circle-mlir/lib/pass/src/ops/ConvOp.h

```diff
@@ -0,0 +1,209 @@
+                                      mlir::ConversionPatternRewriter &rewriter) const override
+  {
+    assert(op.verify().succeeded());
+
+    auto op_name = GetOperationName(op.getOperation());
+
+    LLVM_DEBUG({ llvm::dbgs() << "ConvConv name: " << op_name << "\n"; });
+    LLVM_DEBUG({ llvm::dbgs() << "ConvConv auto_pad: " << op.getAutoPad() << "\n"; });
+    LLVM_DEBUG({ llvm::dbgs() << "ConvConv dilations: " << op.getDilations() << "\n"; });
+    LLVM_DEBUG({ llvm::dbgs() << "ConvConv group: " << op.getGroup() << "\n"; });
+    LLVM_DEBUG({ llvm::dbgs() << "ConvConv kernel_shape: " << op.getKernelShape() << "\n"; });
+    LLVM_DEBUG({ llvm::dbgs() << "ConvConv pads: " << op.getPads() << "\n"; });
+    LLVM_DEBUG({ llvm::dbgs() << "ConvConv strides: " << op.getStrides() << "\n"; });
+
+    if (notYetImplemented(op))
+      return mlir::failure();
+
+    mlir::Value input = adaptor.getX();
+    mlir::Value filter = adaptor.getW();
+    mlir::Value bias = adaptor.getB();
+    bool biasNone = bias.getType().isa<mlir::NoneType>();
+
+    mlir::Location opLoc = op->getLoc();
+
+    // TODO support other ranks for I/O
+
+    mlir::RankedTensorType intype = input.getType().dyn_cast_or_null<mlir::RankedTensorType>();
+    LLVM_DEBUG({ llvm::dbgs() << "ConvConv intype: " << intype << "\n"; });
+    if (intype.getRank() != 4)
+      assert(false);
+    CHECK_VALID_RANK_4(intype);
+
+    mlir::RankedTensorType outtype = op.getType().dyn_cast_or_null<mlir::RankedTensorType>();
+    LLVM_DEBUG({ llvm::dbgs() << "ConvConv outtype: " << outtype << "\n"; });
+    if (outtype.getRank() != 4)
+      assert(false);
+    CHECK_VALID_RANK_4(outtype);
+
+    mlir::Value inputPreTr = input;
+    // for op.pads != [0,0,0,0]
+    std::vector<int32_t> padsValue;
+    if (GetPads(op.getPads(), padsValue))
+    {
+      mlir::Type i32 = rewriter.getI32Type();
+      mlir::RankedTensorType ptype = RankedTensorType::get({4, 2}, i32);
+      llvm::SmallVector<int32_t, 8> pvalue = {
+        0, 0, 0, 0, padsValue[0], padsValue[2], padsValue[1], padsValue[3]};
+      mlir::Location padsval_loc = mlir::NameLoc::get(rewriter.getStringAttr(op_name + "/padsval"));
+      mlir::Value paddings =
+        rewriter.create<ConstOp>(padsval_loc, DenseIntElementsAttr::get(ptype, pvalue));
+
+      // calc output type+shape of Pad
+      auto shape = intype.getShape();
+      assert(shape.size() == 4);
+      int64_t padH = 0, padW = 0;
+      // NOTE if input is unknown, set padH, padW as unknown.
+      // these will be resolved in shape inference.
+      if (shape[2] == std::numeric_limits<int64_t>::min())
+        padH = std::numeric_limits<int64_t>::min();
+      else
+        padH = shape[2] + padsValue[0] + padsValue[2];
+      if (shape[3] == std::numeric_limits<int64_t>::min())
+        padW = std::numeric_limits<int64_t>::min();
+      else
+        padW = shape[3] + padsValue[1] + padsValue[3];
```

Comment:
It looks like the two "if statements" are connected.
How about?
```suggestion
      padH = (shape[2] == std::numeric_limits<int64_t>::min() ? \
              shape[2] : shape[2] + padsValue[0] + padsValue[2]);
      padW = (shape[3] == std::numeric_limits<int64_t>::min() ? \
              shape[3] : shape[3] + padsValue[1] + padsValue[3]);
```